Cython/Compiler/ExprNodes.py

```diff
@@ -7373,7 +7382,11 @@ def generate_result_code(self, code):
                         '%s'
                     '}' % (self.result(), code.error_goto(self.pos)))
         elif self.entry.is_cpp_optional and self.initialized_check:
-            unbound_check_code = self.type.cpp_optional_check_for_null_code(self.result())
+            undereferenced_result = self.result()
+            if self.entry and self.entry.is_cpp_optional and not self.is_target:
+                assert undereferenced_result.startswith("(*") and undereferenced_result.endswith(")")
+                undereferenced_result = undereferenced_result[2:-1]
```

Comment:
Hmm. I don't know how to do this better right now, but it's probably not something that you're proud of either.

Reply:
To me, it feels more like calling `.result()` isn't the right thing and there should be some intermediate method. The `result()` method is clearly doing too much here. Adding an option to it would rather make it worse.

I could imagine splitting most of `calculate_result_code()` out into a new method `calculate_access_code()`, and then calling that from `calculate_result_code()` and from here.